ykcompile/src/lib.rs

```diff
@@ -860,35 +460,35 @@ impl<TT> TraceCompiler<TT> {
             .collect::<Vec<u8>>();
 
         for arg in args {
+            // In which register will this argument be passed?
             // `unwrap()` must succeed, as we checked there are no more than 6 args above.
             let arg_reg = arg_regs.pop().unwrap();
 
-            match arg {
-                Operand::Place(place) => {
-                    // Load argument back from the stack.
-                    let (loc, _) = self.place_to_location(place, false);
-                    match &loc {
-                        Location::Register(reg) => {
-                            let off = stack_index(*reg) * 8;
-                            dynasm!(self.asm
-                                ; mov Rq(arg_reg), [rsp + off]
-                            );
-                        }
-                        Location::Mem(ro) => {
-                            dynasm!(self.asm
-                                ; mov Rq(arg_reg), [Rq(ro.reg) + ro.offs]
-                            );
-                        }
-                        Location::Addr(_) | Location::NotLive => unreachable!(),
-                    };
-                    self.free_if_temp(loc);
+            // Now load the argument into the correct argument register.
+            match self.iplace_to_location(arg) {
+                Location::Register(reg) => {
+                    // The value *was* in a register before we pushed it with caller_save().
+                    // We load it back from the stack now.
+                    // FIXME assumes that anything in a register was pushed. In reality only caller
```

Comment:
I'm struggling to parse the `FIXME`.

Reply:
We pushed all of the caller save registers and the code is assuming that when loading argument registers we will be taking them off the stack. In reality the argument may not have been in a caller save register in the first place.

If that makes sense, I'll paste the above in a comment :P

Reply:
That makes more sense. I think the "code is assuming" bit needs mild rephrasing, but otherwise it's an improvement!